### PR TITLE
add CIP-0094 support

### PR DESCRIPTION
**Pull request by gufmar**
## Description
add pool > vote command
- check for xdd support
- requires Koios API
- shows Poll questions currently open for pool answers
- generates CBOR answer metadata
- online and offline mode

## Motivation and context
add CIP-0094 poll answer support

## How has this been tested?
on and offline mode on preprod

### CODE REVIEW

File: scripts/cnode-helper-scripts/cntools.sh [hunk 1]
```diff
@@ -3065,6 +3067,238 @@ function main {
...
+                cborStr+=$(to_cbor "unsigned" 2) #unsigned 2
+                cborStr+=$(to_cbor "bytes" "${questionHash}") #bytearray of the blake2b-256 hash of the question cbor
+                cborStr+=$(to_cbor "unsigned" 3) #unsigned 3
+                cborStr+=$(to_cbor "unsigned" ${answer}) #unsigned - answer index
+                #CBOR Answer is ready, write it out to disc
+                cborFile="${TMP_DIR}/CIP-0094_${poll_txId}_answer.cbor"
+                #echo -ne "Writing '${cborFile}' to disc ... "
+                xxd -r -ps <<< ${cborStr} 2> /dev/null > ${cborFile}
+                if [ $? -ne 0 ]; then echo -e "\n\n${FG_RED}ERROR, could not write to file!\n\n${NC}"; exit 1; fi
+                if ! submitPoll "${pool_name}"; then
```
Comment (Scitz0) [inline scripts/cnode-helper-scripts/cntools.sh:3292]: See above comment, no need to pass ${pool_name} to funcion

**Review by Scitz0 — APPROVED** [review]
Just some minor non-breaking notes I saw while scrolling through the code. I have yet to test it but will do so next. Seems alright to me. I see that label 674 message is not included in this version. But not criticial I guess.

File: scripts/cnode-helper-scripts/cntools.sh [hunk 2]
```diff
@@ -3065,6 +3067,238 @@ function main {
...
+                println ERROR "\n${FG_RED}ERROR${NC}: no funds available for wallet ${FG_GREEN}${wallet_name}${NC}"
+                waitForInput && continue
+              fi
+              epoch=$(getEpoch)
+              echo 
+              echo "Current ${NETWORK_NAME} epoch: ${epoch}"
+              NETWORK_NAME_LOWER=$(echo "$NETWORK_NAME" | awk '{print tolower($0)}')
+              println LOG "Query ${NETWORK_NAME} polls ..."
+              println ACTION "curl -sSL -f -H \"Content-Type: application/json\" ${CIP0094_POLL_URL}"
+              ! polls=$(curl -sSL -f -H "Content-Type: application/json"  "${CIP0094_POLL_URL}" | jq -r .networks.${NETWORK_NAME_LOWER} 2>&1) && error_msg=${polls} && waitForInput && continue
```
Comment (Scitz0) [inline scripts/cnode-helper-scripts/cntools.sh:3175]: Minor thing but really no need to put polls into error_msg variable if you are not going to print it. Can just skip this extra `&& error_msg=${polls}`

File: scripts/cnode-helper-scripts/cntools.sh [hunk 3]
```diff
@@ -3065,6 +3067,238 @@ function main {
...
+              poll_index=$(echo $polls | jq '. | length')
+              if [[ "$poll_index" -gt 0 ]]; then
+                poll_index_act=0
+                poll_index_cnt=0
+                declare poll_index_txIds=()
+                declare poll_index_titles=()
+                echo "Polls currently open for Pool answers:"
+                while read poll; do
+                  poll_index_cnt=$((poll_index_cnt+1))
+                  if [[ "$epoch" -ge "$(jq '.epoch_cast' <<< $poll)" ]] && [[ "$epoch" -lt "$(jq '.epoch_delegation' <<< $poll)" ]]; then
```
Comment (Scitz0) [inline scripts/cnode-helper-scripts/cntools.sh:3185]: Initial `if [[ "$poll_index" -gt 0 ]]; then` check is good but a second check is needed after filtering out invalid polls and break out as list could be empty after this.

**Review by Scitz0 — CHANGES_REQUESTED** [review]
Changes needed to poll loop as it might render an empty poll list and be stuck in a for-ever loop

File: scripts/cnode-helper-scripts/cntools.sh (hunk 3 above)
Comment (gufmar) [reply scripts/cnode-helper-scripts/cntools.sh:3106]: added this second check, and reordered the whole steps. 
This check for active polls is now the first step. only if there are active polls, it comes to the selection of pool, payment wallet and the poll answer.

**Comment by Scitz0** [comment]
Looks good but will wait with approval until I have tested it. Also if you want to add label 674 before merging.

**Review by Scitz0 — APPROVED** [review]
Tested with hybrid / offline flow and it worked great, approving PR but leaving unmerged as mentioned for possible addition of 674 message.

**Comment by gufmar** [comment]
last commit adds support for label 674 messages

**Pull request merged**